lib/Common/DataStructures/ClusterList.h

```diff
@@ -0,0 +1,402 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#pragma once
+
+#include "../Core/Assertions.h"
+#include "../Memory/Allocator.h"
+
+/*
+ * ClusterLists are intended to solve the problem of determining what connected set
+ * groups are present in a graph while iterating over a list of connections.
+ */
+
+template<typename indexType, class TAllocator>
+class ClusterList
+{
+private:
+    // The maximum index enabled; basically an array bounds for the set.
+    indexType maxIndex;
+    // The low-level data store backing the structure. This is a list initialized as the
+    // identity mapping (list[i] = i), and is then updated as links are added so that an
+    // entry in the list is the index of an entry of equal or lower number which is part
+    // of the same set. This is later updated to be the index of the element with lowest
+    // index in the set for all entries, avoiding situations where lookups would need to
+    // follow a chain.
+    indexType* list;
+#if DBG
+    // During the merge process, we may end up with some situations where set membership
+    // is chained; that is, that node 7 is in set 5, but 5 is in set 4. Following all of
+    // the merge operations we have a step to reconcile all of this information and then
+    // update each element of the list to point to the least member of its set. This can
+    // be checked in debug builds to see if we've done this step or not.
+    bool consolidated = false;
+#endif
+public:
+    // Create a new ClusterList with a specified maximum node number
+    ClusterList(indexType maxCount)
+        : maxIndex(maxCount)
+        , list(nullptr)
+#if DBG
+        , consolidated(true)
+#endif
+    {
+        // Note that if you want to use this with something other than an ArenaAllocator, then
+        // you'll have to add clean-up mechanics
+        list = AllocatorNewArrayLeaf(TAllocator, this->alloc, indexType, maxCount);
```

Comment:
ClusterList has no alloc member.

Reply:
Woops, fixed; that's a problem with replicating changes to a sibling templated type that's not instantiated.